### PR TITLE
PHP DB code: clean up the logic, and allow for > 1 readonly replica

### PR DESCRIPTION
**Pull request by davidpanderson**
Also always use mysqli interface

### CODE REVIEW

**Comment by brevilo** [comment]
> Also always use mysqli interface

Please, use atomic commits so that one can review separate topics separately.

File: html/inc/boinc_db.inc [hunk 1]
```diff
@@ -27,119 +27,97 @@ incs();
...
+        $passwd = parse_config($config, '<db_passwd>');
+        // OK if host is null; it will use localhost
+        if (!$name || !$user || !$passwd) {
+            error_log("BoincDb::get_aux(): must specify DB name, user, passwd");
+            return;
+        }
+        $retval = $instance->init_conn($user, $passwd, $host, $name);
+        if ($retval) {
+            if ($instance->do_query("use $name")) {
+                //error_log("BoincDb::get_aux(): connected to DB $dbnum");
```
Comment (brevilo) [inline html/inc/boinc_db.inc:81]: Can be removed.

File: html/inc/boinc_db.inc [hunk 2]
```diff
@@ -27,119 +27,97 @@ incs();
...
+            $passwd = parse_config($config, sprintf('<replica%s_db_passwd>', $r));
+            if ($host && $name && $user && $passwd) {
+                $retval = $instance->init_conn($user, $passwd, $host, $name);
                 if ($retval) {
-                    // needed for places where we do direct queries
-                    if (!$instance->do_query("use $name")) {
-                        error_log("BoincDb::get_aux(): Couldn't select database $name on $replica_host (giving up)");
-                        $instance = null;
+                    if ($instance->do_query("use $name")) {
+                        //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
```
Comment (brevilo) [inline html/inc/boinc_db.inc:61]: Can be removed.

File: html/inc/boinc_db.inc [hunk 3]
```diff
@@ -27,119 +27,97 @@ incs();
...
-        } else {
-            // needed for places where we do direct queries
-            if (!$instance->do_query("use $name")) {
-                error_log("BoincDb::get_aux(): Couldn't select database $name on $host (giving up)");
-                $instance = null;
+        $host = parse_config($config, '<db_host>');
+        $user = parse_config($config, '<db_user>');
+        $name = parse_config($config, '<db_name>');
+        $passwd = parse_config($config, '<db_passwd>');
+        // OK if host is null; it will use localhost
```
Comment (brevilo) [inline html/inc/boinc_db.inc:73]: How? I don't see it being set explicitly. Do you depend on `mysqli` to do that? If so, it [falls back](https://www.php.net/manual/en/mysqli.construct.php) to use [`mysqli.default_host`](https://www.php.net/manual/en/mysqli.configuration.php#ini.mysqli.default-host) which doesn't have a [default value](https://www.php.net/manual/en/mysqli.configuration.php) configured, and I can't find one explicitly defined here either.

**Comment by brevilo** [comment]
Apart from the above I like the approach. Flexible while not overly complex.

**Comment by davidpanderson** [comment]
I removed the 'use db_name' query (not needed with mysqli)
and made the default 'localhost' explicit.

File: html/inc/boinc_db.inc (hunk 3 above)
Comment (brevilo) [reply html/inc/boinc_db.inc]: Thanks

File: html/inc/boinc_db.inc [hunk 4]
```diff
@@ -57,31 +57,27 @@ class BoincDb extends DbConn {
             if ($host && $name && $user && $passwd) {
                 $retval = $instance->init_conn($user, $passwd, $host, $name);
                 if ($retval) {
-                    if ($instance->do_query("use $name")) {
-                        //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
-                        self::$instance = $instance;
-                        return;
-                    }
+                    //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
```
Comment (brevilo) [inline html/inc/boinc_db.inc:60]: Please remove dead code like this. You can add it back for debugging if needed. Alternatively, use pre-processor macros.

File: html/inc/boinc_db.inc [hunk 5]
```diff
@@ -57,31 +57,27 @@ class BoincDb extends DbConn {
...
             return;
         }
         $retval = $instance->init_conn($user, $passwd, $host, $name);
         if ($retval) {
-            if ($instance->do_query("use $name")) {
-                //error_log("BoincDb::get_aux(): connected to DB $dbnum");
-                self::$instance = $instance;
-                return;
-            }
+            //error_log("BoincDb::get_aux(): connected to DB $dbnum");
```
Comment (brevilo) [inline html/inc/boinc_db.inc:78]: Please remove dead code like this. You can add it back for debugging if needed. Alternatively, use pre-processor macros.

**Comment by brevilo** [comment]
@bema-aei should ultimately review this as well.

File: html/inc/boinc_db.inc [hunk 6]
```diff
@@ -27,119 +27,93 @@ incs();
...
-                }
-            }
-            self::get_aux($readonly, $fallback_mode);
-            if (!self::$instance) {
-                if ($generating_xml) {
-                    xml_error(-138, "Can't connect to database");
-                } else {
-                    error_page("Can't connect to database");
-                }
+        if (isset(self::$instance)) {
```
Comment (brevilo) [inline html/inc/boinc_db.inc:98]: Hm, this implies that there can only be a single instance at a time. What we need here is a hash of instances (indexed by `$dbnum`) such that multiple instances are retained and/or can be switched between. E.g. the server status page needs to query different databases (here: replicas).

Comment (bema-aei) [reply html/inc/boinc_db.inc:98]: I think a hash of instances might be a bit of an overkill, I think E@H is the only project that uses as much as two. However a requirement for us would be to access different replicas at least in sequence. I did this in d1f99043877552ea48ac2912dc4ec9ac9a1efddd

**Comment by bema-aei** [comment]
This works fro us in principle, with the change I mentioned above. I updated my original branch with your changes, with two important changes:
- including a change that affects the whole web code ("use mysqli everywhere") in another commit with very limited scope and not even mentioning it in the commit message is just asking for trouble. It's difficult to review, difficult to test. Worst of all,  if someone of some project later finds it to cause trouble for him, he is unlikely to ever find the commit that breaks his project. So I split your commit in two (leaving the author intact).
- I added the commit I mentioned above that is required to make this work for us. It allows to address multiple DB hosts in sequence without much effort.
- I also changed the debug output to suit my needs, then reverted it. You can re-add it by reverting the 'Revert' commit if needed.
Please have a look at that branch.

File: html/inc/boinc_db.inc (hunk 6 above)
Comment (brevilo) [reply html/inc/boinc_db.inc:98]: Ok, but please add a comment that states that there's only ever be one current DB instance at a time. IOW, the current instance gets created, reused or replaced by a new one.

File: html/inc/boinc_db.inc (hunk 4 above)
Comment (brevilo) [reply html/inc/boinc_db.inc:60]: Fixed by @bema-aei

File: html/inc/boinc_db.inc (hunk 5 above)
Comment (brevilo) [reply html/inc/boinc_db.inc:78]: Fixed by @bema-aei

File: html/inc/boinc_db.inc (hunk 6 above)
Comment (bema-aei) [reply html/inc/boinc_db.inc:98]: Is this a requirement?

Comment (brevilo) [reply html/inc/boinc_db.inc:98]: Well no, but that's how you implemented it, isn't it?

Comment (brevilo) [reply html/inc/boinc_db.inc:98]: Resolved [here](https://github.com/BOINC/boinc/pull/5695#discussion_r1698695506)

**Comment by AenBleidd** [comment]
@bema-aei, @brevilo, are you ok with this PR? Does it need anything else (like testing on your side) or we can move forward and merge it?

**Comment by AenBleidd** [comment]
@bema-aei, @brevilo, could you please take a look at this PR and test it?
Thank you in advance.

**Comment by bema-aei** [comment]
First of all I like the idea of allowing an arbitrary number of replicas. And thank you for implementing a close() function.

[This code](https://github.com/BOINC/boinc/blob/4b4ba9dbf83384f272f083875cb7f187bc6f6f67/html/inc/boinc_db.inc#L107-L109) in get() is still unchanged, i.e. if $instance is set, get() will return the same instance even if $dbnum is different.

At least this doesn't correspond to what I proposed in my branch. Is the "user" of this code required to close one connection / instance before opening another? If so, it should be noted somewhere, I didn't expect that.

**Comment by bema-aei** [comment]
Also too I there is an uncaught exception if the DB connection is refused for some reason. I get:
```
PHP Fatal error:  Uncaught mysqli_sql_exception: Connection refused in /srv/BOINC/live-webcode-test/html/inc/db_conn.inc:37
Stack trace:
#0 /srv/BOINC/live-webcode-test/html/inc/db_conn.inc(37): mysqli->__construct()
#1 /srv/BOINC/live-webcode-test/html/inc/boinc_db.inc(65): DbConn->init_conn()
#2 /srv/BOINC/live-webcode-test/html/inc/boinc_db.inc(117): BoincDb::get_aux()
#3 /srv/BOINC/live-webcode-test/html/ops/eah_server_status.php(681): BoincDb::get()
#4 /srv/BOINC/live-webcode-test/html/ops/eah_server_status.php(1108): seconds_behind_master()
#5 {main}
  thrown in /srv/BOINC/live-webcode-test/html/inc/db_conn.inc on line 37
```

**Comment by davidpanderson** [comment]
- I added the check for different dbnum in get()
- Yes, you can only use on DB at a time
- An exception would be thrown if had previously called
  mysqli_report(MYSQLI_REPORT_STRICT);
  presumably because you want to handle exceptions.
  But in any case I added a try/catch.

**Comment by bema-aei** [comment]
Thanks, this works for us technically. We do use a recent MariaDB, which has some defaults different from MySQL, hence the exception.

However, for reasons I pointed out already I can't really approve the first commit which combines different features w/o even mentioning in the commit message, so I can't recommend to merge this branch as it is.

**Comment by brevilo** [comment]
> We do use a recent MariaDB, which has some defaults different from MySQL, hence the exception.

How does a PHP module's config relate to the MySQL flavor used as the DB backend?

**Comment by bema-aei** [comment]
> > We do use a recent MariaDB, which has some defaults different from MySQL, hence the exception.
> 
> How does a PHP module's (`mysqli`) config relate to the MySQL flavor used as the DB backend?

Oh. Sorry, was mislead. Indeed, that's the mysqli config.

**Pull request merged**